File: GRASS.py
# -*- coding: utf-8 -*-
#
#  GRAS Stemmer
#  Authors:   Federico Ghirardelli, Marco Romanelli
#  A.A.:      2016-2017
#

import getopt, sys
from os import path
from getopt import GetoptError, getopt
from collections import Counter
from igraph import *
from operator import itemgetter
from itertools import combinations, takewhile
import cPickle as pickle
from numpy import mean

# Set the interpreter's 'check interval', how often to perform periodic checks
sys.setcheckinterval = 1000

##############################################################
###############   FUNCTIONS AND SUBROUTINES   ################
##############################################################

# Compute Longest Common Prefix (index)
def lcp(x, y):
	# len(x) must be shorter than len(y)
	if not x or not y:
		return 0

	i = 0
	while ((i < len(x)) and (x[i] == y[i])): i += 1

	return i

# Compute Longest Common Suffix (pair of substrings)
def lcs(x, y):
	i = 0
	while ((i < len(x)) and (x[i] == y[i])): i += 1

	return (x[i:], y[i:])

# Compute cohesion function as described in the paper
def cohesion(p_neighbors, v_neighbors):
	return (1 + len(set(p_neighbors).intersection(v_neighbors))) / len(v_neighbors)

# Procedure for parsing the input (lexicon) file
def lexicon_parser(path, readnumbers=True, lower_bound=0, upper_bound=None):
	# assuming path is a regular file path
	lexicon = []

	fp = open(lexicon_path, "r")
	signature = [fp.readline() for i in range(0,2)]

	# verify terrier lexicon's "signature"
	# that means, check if it's a real terrier lexicon file
	if not ((signature[0][:23] == "Setting TERRIER_HOME to") and (signature[1][:20] == "Setting JAVA_HOME to")):
		print "- Error while parsing the lexicon."
		return [], 0

	# for every line
	for w in fp:
		try:
			word = w[:w.index(',')].strip()
		except:
			print "Oops!  word wrong at line %d" % pos		# if there's no ',' char
			sys.exit()
		if word[0].isdigit() and not readnumbers:
			continue
		lexicon.append(word.decode("utf-8"))

	fp.close()

	# cut lexicon with lower/upper bounds
	if lower_bound != 0:
		lexicon = lexicon[lower_bound:]
	if upper_bound != None:
		lexicon = lexicon[:(upper_bound-lower_bound)]

	average_length = int(mean([len(w) for w in lexicon]))

	return lexicon, average_length

# Show an help message
def help_message():
	print "Usage: python GRASS.py -l <lexicon-path> [OPTIONS]\n"
	sys.exit()


##############################################################
###############  			 MAIN   		  ################
##############################################################

# Default stemmer parameters
l = None
l_forced = False				# if True l will be computed as the average word length
alpha = 4
delta = 0.8
readnumbers = True

# other "global" variables
freq_filename_dump = "suffix_pair_freq_l-p1.txt"
freq_filename_load = None

# Output file path
stems_file_path = "stems.txt"

# Command line arguments parsing
if len(sys.argv) < 2:
	help_message()

# Overriding value if passed as argument in command line
lexicon_path = None
try:
	opts, args = getopt(sys.argv[1:], "l:", ["alpha=", "delta=", "l=", "no-numbers", "freq-file="])
except GetoptError:
	    help_message()
for opt, arg in opts:
	if opt == "-l":
		if not ((type(arg) is str) and path.isfile(arg)):
			print "- Bad lexicon file."
			sys.exit()
		lexicon_path = arg.strip()
	if opt == '--alpha':
		alpha = int(arg)
	elif opt == '--l':
		l = int(arg)
		l_forced = True
	elif opt == '--delta':
		delta = float(arg)
	elif opt == '--no-numbers':
		readnumbers = False
		print "+ [OPTION] Do not parse numbers."
	elif opt == '--freq-file':
		freq_filename_load = str(arg)
		print "+ [OPTION] Suffix pair frequencies file selected."

if lexicon_path is None:
	print "- Missing lexicon file."
	sys.exit()

# Read input lexicon
print "+ Parsing lexicon..."
lexicon, average_length = lexicon_parser(lexicon_path, readnumbers)

if len(lexicon) == 0:
	print "- Error while parsing the lexicon."
	sys.exit()

lexicon.sort() 			# sort lexicon

print "+ Lexicon parsed (%d words)." % len(lexicon)

# updating l value if l_forced=False
if not l_forced:
	l = average_length
del average_length

print "+ Parameters:"
print "\t+ l = %d" % l
print "\t+ alpha = %d" % alpha
print "\t+ delta = %.1f" % delta

print "+ Clustering words in classes..."
classes = [[]]

# clustering words into classes
i = 0
j = 0

while i < len(lexicon):
	# Compute lcp between every two consecutive words
	# and put them in the same class if lcp >= l
	w1 = lexicon[i]
	w2 = lexicon[i+1] if (i < len(lexicon)-1) else ""
	classes[j].append(w1)
	if lcp(w1, w2) < l:
		classes.append([])
		j += 1
	i += 1

del classes[j]	# empty class
print "+ Done (%d word classes created)." % len(classes)

# If no freq file found compute it:
if freq_filename_load == None:
###################### Algorithm #1 ############################
	print "+ Computing alpha-frequencies..."
	frequencies_temp = Counter({})
	append = list.append				# speed-up
	i = 0

	# loop for every class
	for m in range(0, len(classes)):
		suffix_array = []

		if len(classes[m]) == 1:
			continue

		# computing pairs of suffixes (current class)
		for w1, w2 in combinations(classes[m], 2):
			append(suffix_array, lcs(w1, w2))

		# combining two counters, local and global (so far)
		frequencies_temp = frequencies_temp + Counter(suffix_array)

	print "+ %d \"raw\" suffixes pair found." % len(frequencies_temp)

	# Storing "raw" suffix pair frequencies for further uses
	freq_filename_dump = freq_filename_dump.replace("p1", str(l))
	pickle.dump(frequencies_temp, open(freq_filename_dump, "wb"))

	# Removing names (clear memory)
	del suffix_array

else:
	# otherwise load it
	with open(freq_filename_load, "rb") as fp:
		frequencies_temp = pickle.load(fp)

	print "+ Load external \"raw\" suffix pair (%s)." % len(frequencies_temp)

# Removing suffixes with frequency less than alpha
print "\t+ Doing some improvement..."
frequencies = {k:v for k,v in frequencies_temp.iteritems() if v >= alpha}

print "+ Done: %d alpha-suffixes found." % len(frequencies)
del frequencies_temp

print "+ Generating graph G=(V,E) ..."
g = Graph(directed=False)

# adding a vertex for each word in the lexicon
for i in range(0, len(lexicon)):
	g.add_vertex(lexicon[i])

print "\t+ |V|=%d" % g.vcount()

# compute edges
for m in range(0, len(classes)):
	if len(classes[m]) == 1:
		continue

	# computing pairs of suffixes (current class)
	alpha_suffix_array = [(w1, w2) for w1, w2 in combinations(classes[m], 2) if lcs(w1, w2) in frequencies]

	# if there's a suffix pair with frequency greater than alpha
	# induced by the two words/nodes, then add an edge
	for w1, w2 in alpha_suffix_array:
		g.add_edge(w1, w2, weight=frequencies[lcs(w1, w2)])

# if something weird happened
if len(g.es) == 0:
	print "\t- The number of edges is equal to zero. Something wrong?"
	sys.exit()
else:
	print "\t+ |E|=%d" % g.ecount()

# Remove names (clear memory)
del alpha_suffix_array
del classes
classes = []

###################### Algorithm #2 ############################
print "+ Identifying classes..."
jj = 0
early_quitting = False

# while the graph is not empty
while (g.vcount() != 0) and not early_quitting:
	S = []

	degree_list = g.degree() # degree list for each node
	index, value = max(enumerate(degree_list), key=operator.itemgetter(1)) # index-value of maximum degree node
	u = g.vs[index] 			# = pivot

	S.append(index)				# S={u}

	# find the adjacency list for u and v
	u_adjacency = g.neighbors(u)
	u_dec_adjacency = []
	for v in u_adjacency:
		edge_id = g.get_eid(u, v)
		w = g.es[edge_id]["weight"]
		u_dec_adjacency.append((v, w))

	u_dec_adjacency.sort(key=operator.itemgetter(1), reverse=True)	# sort by decreasing weight

	# append node v to the class if choesion value is greater than delta
	for (v, weight) in u_dec_adjacency:
		if cohesion(u_adjacency,g.neighbors(v)) >= delta:
			S.append(v)
		else:
			g.delete_edges(g.get_eid(u, v))

	# output class S
	classes.append([g.vs[v]["name"] for v in S])

	# Remove from G vertexes which are in S and adjacent edges
	g.delete_vertices(S)

	# loop breaking
	if len(g.es) == 0:
		early_quitting = True

# when quitting with early_quitting=True there're still vertex in the graph to be added in some class(es)
# I assume that there'd go in singleton sets
if early_quitting:
	print "\t+ Adding additional %d singletons." % len(g.vs)
	classes += [[v["name"]] for v in g.vs]

# remove names (clear memory)
del g
print "+ Classed created."

print "+ Storing items..."
output = []

# create a word-stem list in order to sort it later
for c in classes:
	if len(c) == 1:
		output.append((c[0], c[0]))
	else:
		for word in c[1:]:
			output.append((word, c[0]))

output.sort(key=operator.itemgetter(0))		# ordino le parole del dizionario (come in origine)

# store the word - stem pairs in the lookup file
fp = open(stems_file_path, "w")
for word, stem in output:
	s = "%s\t%s\n" % (word, stem)
	fp.write(s.encode("utf-8"))

fp.close()

print "+ Done. Bye."
sys.exit()

File: GRASS.java
/*
  GRAS Stemmer (Java version)
  Authors:   Federico Ghirardelli, Marco Romanelli
  A.A.:      2016-2017
*/

import java.io.File;
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.BufferedWriter;
import java.io.FileWriter;
import java.util.Properties;
import java.io.UnsupportedEncodingException;
import java.util.HashMap;
import java.util.Map;
import java.util.TreeMap;
import java.util.ArrayList;
import java.util.List;
import java.util.Set;
import java.util.Arrays;
import java.util.HashSet;
import java.util.Collections;
import java.util.Iterator;

// JGraphT graph library (jgrapht-core-1.0.1.jar)
import org.jgrapht.*;
import org.jgrapht.graph.*;
import org.jgrapht.Graph;
import org.jgrapht.alg.util.VertexDegreeComparator;
import org.jgrapht.alg.NeighborIndex;

public class GRASS
{
	private static String PROP_FILE = "config.properties";
	private static String stem_filename = "stem.txt";
	private static int l;
	private static int alpha;
	private static double delta;
	private static Boolean computeValuel;

	public static void main(String[] args) throws Exception
	{
		String lexicon_filename;
		ArrayList<String> lexicon;
		ArrayList<ArrayList<String>> wordClasses;
		ArrayList<ArrayList<String>> stemClasses;
		HashMap<ArrayList<String>, Integer> suffixes;
		SimpleWeightedGraph<String, DefaultWeightedEdge> graph;

		if (args == null || args.length != 1) {
			System.out.println("! Missing lexicon.");
			System.exit(0);
		}
		lexicon_filename = args[0];

		// Welcome message
		System.out.println("GRAS Stemmer (Java version)");
		System.out.println("Authors: Marco Romanelli, Federico Ghirardelli\n");

		// Load parameters from config file
		loadProperties();

		// read words from lexicon file
		System.out.println("+ Parsing lexicon file...");
		lexicon = readLexicon(lexicon_filename);
		System.out.printf("+ %d words found.\n", lexicon.size());
		if (computeValuel)	System.out.printf("\t+ l = %d (average word length)\n.", l);

		// sort lexicon (alphabetically)
		System.out.format("+ Sorting lexicon... (%d)\n", lexicon.size());
		Collections.sort(lexicon);

		// create the set of classes s.t. any two words in any class have the lcp of length > 3
		wordClasses = createWordClasses(lexicon, l);
		System.out.format("+ Classes created (%d).\n", wordClasses.size());

		// compute alpha frequencies
		System.out.println("+ Computing suffix alpha frequencies...");
		suffixes = computeSuffixesFrequencies(wordClasses);
		System.out.format("+ Done. (%d).\n", suffixes.size());

		// Generate graph
		System.out.println("+ Generating graph...");
		graph = generateGraph(wordClasses, suffixes);
		System.out.format("+ Done. (|V|=%d, |E|=%d)\n", graph.vertexSet().size(), graph.edgeSet().size());

		// "delete" unused objects
		lexicon = null;
		wordClasses = null;

		// Identify classes
		System.out.println("+ Identifying classes...");
		stemClasses = algorithm2(graph);
		System.out.format("+ Done. (%d).\n", stemClasses.size());

		// "delete" unused ojbects
		graph = null;

		// Writing to file...
		stem_filename += stem_filename.format("_%s_%s_0%s.txt", l, alpha, (int)(delta*10));
		System.out.printf("+ Writing stems to file \"%s\"\n", stem_filename);
		storeStems(stemClasses, stem_filename);
		System.out.println("+ Stems stored!");
		System.out.println("+ Bye!\n");
	}

	public static void loadProperties()
	{
		Properties prop = new Properties();
		FileInputStream input = null;

		try
		{
			input = new FileInputStream(PROP_FILE);

			// load a properties file
			prop.load(input);

			// get the property values
			l = Integer.parseInt(prop.getProperty("l"));
			alpha = Integer.parseInt(prop.getProperty("alpha"));
			delta = Double.parseDouble(prop.getProperty("delta"));
			computeValuel = prop.getProperty("auto-compute-l").equals("True") ? true : false;
			stem_filename = prop.getProperty("output-file");

			System.out.println("+ Parsing config.properties...");
			if (computeValuel)
				System.out.println("\t+ l will set to word avarage length.");
			else
				System.out.printf("\t+ l = %d (forced to this value)\n", l);

			System.out.printf("\t+ alpha = %d\n\t+ delta = %.2f\n", alpha, delta);

		}
		catch (IOException e)
		{
			System.out.format("! Error loading config.properties (%s).\n", e);
			System.exit(1);
		}
		finally
		{
			if (input != null) {
				try
				{
					input.close();
				}
				catch  (IOException e)
				{
					System.out.format("! Error loading config.properties (%s).\n", e);
				}
			}
		}
	}

	public static ArrayList<String> readLexicon(String filename) throws Exception
	{
		ArrayList<String> filedata = null;

		try
		{
			filedata = new ArrayList<String>();
			File fp = new File(filename);
			BufferedReader in = new BufferedReader(new InputStreamReader(new FileInputStream(fp), "UTF-8"));

			String line = null;
			String word = null;

			int i = 0;
			// Skip the first 2 lines
			while ((line = in.readLine()) != null && (i++ < 1)) continue;

			while ((line = in.readLine()) != null) {
				word = line.split(",")[0];
				filedata.add(word);
			}
		}
		catch (UnsupportedEncodingException e)
		{
			System.out.println("! UTF-8 decoding error: " + e.getMessage());
		}
		catch (IOException e)
		{
			System.out.println("! IO exception error: " + e.getMessage());
		}
		catch(Exception e)
		{
			System.out.println("! Unknown error: "+ e.getMessage());
		}
		finally
		{
			if ((filedata == null) || filedata.isEmpty()) System.exit(2);
		}

		// Compute l value (average word length)
		if (computeValuel)
		{
			int sum = filedata
				.stream()
				.mapToInt(w -> w.length())
				.sum();

			l = sum / filedata.size();
		}

		return filedata;
	}

	public static ArrayList<ArrayList<String>> createWordClasses(ArrayList<String> lexicon, int threshold)
	{
		String w1, w2 = null;
		ArrayList<ArrayList<String>> allClasses = new ArrayList<ArrayList<String>>();
		ArrayList<String> singleClass = new ArrayList<String>();

		int i = 0;
		int lexicon_size = lexicon.size();

		lexicon.add(lexicon_size, "");		// così controlla anche l'ultima parola

		while (i < lexicon_size)
		{
			w1 = lexicon.get(i);
			w2 = lexicon.get(i+1);
			singleClass.add(w1);

			if (longestCommontPrefix(w1, w2) < threshold)
			{
				allClasses.add(singleClass);
				singleClass = new ArrayList<String>();
			}

			i++;
		}

		lexicon.remove(lexicon_size);		// il lexicon torna quello vero

		return allClasses;
	}

	public static HashMap<ArrayList<String>, Integer> computeSuffixesFrequencies(ArrayList<ArrayList<String>> wClasses)
	{
		HashMap<ArrayList<String>, Integer> suffixes = new HashMap<ArrayList<String>, Integer>();
		ArrayList<String> currentClass = new ArrayList<String>();
		ArrayList<String> pair = null;
		String w1, w2, s1, s2;
		int i, j, k, lcp;
		Integer oldPairFreq;

		for (i = 0; i < wClasses.size(); i++)
		{
			currentClass = wClasses.get(i);

			for (k = 0; k < currentClass.size(); k++)
			{
				w1 = currentClass.get(k);
				for (j = 0; j < k; j++)
				{
					w2 = currentClass.get(j);
					lcp = longestCommontPrefix(w2, w1);
					s1 = w1.substring(lcp);
					s2 = w2.substring(lcp);
					pair = new ArrayList<String>();
					pair.add(s1);
					pair.add(s2);

					oldPairFreq = suffixes.put(pair, 1);
					if (oldPairFreq != null)
						suffixes.put(pair, oldPairFreq + 1);
				}
			}
		}

		return suffixes;
	}

	public static SimpleWeightedGraph<String, DefaultWeightedEdge> generateGraph(ArrayList<ArrayList<String>> wordClasses, HashMap<ArrayList<String>, Integer> suffixes)
	{
		SimpleWeightedGraph<String, DefaultWeightedEdge> graph = new SimpleWeightedGraph<String, DefaultWeightedEdge>(DefaultWeightedEdge.class);
		DefaultWeightedEdge edge;
		ArrayList<String> pair = null;
		String w1, w2, s1, s2;
		int lcp;
		Integer pairFreq;

		// adding vertices
		for (ArrayList<String> currentClass : wordClasses)
			for (String word: currentClass)
				graph.addVertex(word);

		System.out.println("+ Vertexes created... ");
		System.out.println("+ Adding edges... ");

		// adding edges (WARNING: bottleneck!)
		for (ArrayList<String> currentClass : wordClasses)
		{
			if (currentClass.size() == 1) continue;

			for (int i = 0; i < currentClass.size(); i++)
			{
				for (int j = 0; j < i; j++)
				{
					w1 = currentClass.get(i);
					w2 = currentClass.get(j);
					lcp = longestCommontPrefix(w2, w1);
					s1 = w1.substring(lcp);
					s2 = w2.substring(lcp);
					pair = new ArrayList<String>();
					pair.add(s1);
					pair.add(s2);

					pairFreq = suffixes.get(pair);
					if ((pairFreq != null) && (pairFreq >= alpha))
					{
						edge = graph.addEdge(w2, w1);
						graph.setEdgeWeight(edge, pairFreq);
					}
				}
			}
		}

		return graph;
	}

	public static ArrayList<ArrayList<String>> algorithm2(SimpleWeightedGraph<String, DefaultWeightedEdge> graph)
	{
		ArrayList<ArrayList<String>> classes = new ArrayList<ArrayList<String>>();
		ArrayList<String> s;
		HashSet<String> adjU, adjV, intersection;
		VertexDegreeComparator comparator = new VertexDegreeComparator(graph, VertexDegreeComparator.Order.ASCENDING);
		String u;
		double chs;

		while (graph.vertexSet().size() > 0)
		{			// u <-- vertex with maximum degree in G
			u = Collections.max(graph.vertexSet(), comparator);
			s = new ArrayList<String>();

			// s = S (current class)
			s.add(u);

			// for all v in Adjacent(u)...
			adjU = new HashSet<String>(Graphs.neighborListOf(graph, u));

			// taken in decrease order of w(u,v)...
			adjU = weightDecreaseOrdering(graph, adjU, u);

			for (String v : adjU)
			{
				adjV = new HashSet<String>(Graphs.neighborListOf(graph, v));
				intersection = new HashSet<String>(adjU);
				intersection.retainAll(adjV);
				chs = cohesion(intersection.size(), adjV.size());

				if (chs > delta)
					s.add(v);
				else
					graph.removeEdge(graph.getEdge(u, v));
			}

			// Store class S
			classes.add(s);

			// From G remove the vertices in S and their incident edges.
			for (String x : s)
			{
				adjV = new HashSet<String>(Graphs.neighborListOf(graph, x));
				for (String y : adjV)
					graph.removeEdge(graph.getEdge(x, y));
				graph.removeVertex(x);
			}

			if (graph.edgeSet().size() == 0)
			{
				System.out.println("+\t Optimizing...");
				for (String v : graph.vertexSet())
					classes.add(new ArrayList<String>(Arrays.asList(v)));
				break;
			}
		}

		return classes;
	}

	public static HashSet<String> weightDecreaseOrdering(SimpleWeightedGraph<String, DefaultWeightedEdge> graph, HashSet<String> adj, String v)
	{
		HashSet<String> rlist = new HashSet<String>();
		HashMap<String, Integer> map = new HashMap<String, Integer>();
		DefaultWeightedEdge edge;
		Iterator iterator;
		int w;

		for (String u : adj)
		{
			edge = graph.getEdge(v, u);
			w = (int) graph.getEdgeWeight(edge);
			map.put(u, w);
		}

		iterator = map.entrySet().iterator();
		while(iterator.hasNext())
		{
			Map.Entry value = (Map.Entry)iterator.next();
			rlist.add((String) value.getKey());
		}

		return rlist;
	}

	public static int longestCommontPrefix(String a, String b)
	{
		if ("".equals(a) || "".equals(b)) return 0;
		if (a.length() > b.length())
		{
			String x = a;
		 	a = b;
		 	b = x;
		}

		int i, a_n;
		a_n = a.length();

		i = 0;
		while ((i < a_n) && (a.charAt(i) == b.charAt(i))) i++;

		return i;
	}

	public static double cohesion(int adjIntersection, int adjV)
	{
		double chs;

		chs = (1 + adjIntersection) / adjV;

		return chs;
	}

	public static void storeStems(ArrayList<ArrayList<String>> classes, String filename)
	{
		ArrayList<String> output = new ArrayList<String>();
		BufferedWriter buffer = null;
		Iterator iter;
		String stem, s;
		FileWriter fw;
		File fp;

		// Sorting words
		for (ArrayList<String> cc : classes)
		{

			stem = cc.get(0);
			if (cc.size() == 1)
				s = stem + "\t" + stem + "\n";
			else
			{
				s = new String();
				iter = cc.listIterator();
				iter.next();
				while (iter.hasNext())
					s += iter.next() + "\t" + stem + "\n";
			}
			output.add(s);
		}
		Collections.sort(output);

		 try
		 {
			 fp = new File(filename);
			 fp.createNewFile();
			 fw = new FileWriter(fp);
			 buffer = new BufferedWriter(fw);

			 for (String line : output)
			 	buffer.write(line);
		 }
		 catch (IOException e)
		 {
			 System.out.format("! IO Error: %s\n", e);
		 }
		 finally
		 {
			 try
			 {
				 if (buffer != null) buffer.close();
			 }
			 catch (IOException e)
			 {
				  System.out.format("! IO Error: %s\n", e);
			 }
		 }
	}
}
